File: 3-go_complete/basic/fast/dp.cpp
#include <iostream>
#include <ctime>
#include <Windows.h>
 
 
float inner_product(const float* x, const float* y, const long & len){
    float prod = 0.0f;
    long i;
    for (i=0;i<len;i++){
        prod+=x[i]*y[i];
    }
    return prod;
}
 
int main(){
    SetConsoleOutputCP(CP_UTF8); // 设置控制台输出为 UTF-8 编码

    const int len2 = 10000000;
    float *crr=new float[len2];
    float *drr=new float[len2];
    for (int i=0;i<len2;i++){
        int value=i%10;
        crr[i]=value;
        drr[i]=value;
    }
 
    float prod;
    clock_t begin;
    clock_t end;
    begin=GetTickCount(); //从操作系统启动经过的毫秒数
    prod=inner_product(crr,drr,len2);
    end=GetTickCount(); //从操作系统启动经过的毫秒数
    std::cout<<"蛮力计算内积 "<<prod<<"\t用时"<<(double)(end-begin)<<"毫秒"<<std::endl;
}

// g++ -O2 basic/fast/dp.cpp -o basic/fast/dp    不开启编译优化，耗时多一倍
// ./basic/fast/dp

File: 3-go_complete/basic/fast/dp.go
package main

import (
	"fmt"
	"time"
)

func InnerProduct(x []float32, y []float32) float32 {
	var rect float32 = 0
	for i := 0; i < len(x); i++ {
		rect += x[i] * y[i]
	}
	return rect
}

func main() {
	const len2 int = 10000000
	crr := []float32{}
	drr := []float32{}
	for i := 0; i < len2; i++ {
		value := float32(i % 10)
		crr = append(crr, value)
		drr = append(drr, value)
	}

	begin := time.Now().UnixMilli()
	prod := InnerProduct(crr, drr)
	end := time.Now().UnixMilli()
	fmt.Printf("蛮力计算内积 %.0f\t\t用时%d毫秒\n", prod, end-begin)
}

// go run basic/fast/dp.go
